### DISCUSSION TITLE
[SDK] Question on the Design Choice of Signer Class in zkSync's ether SDK for MetaMask Integration

### DISCUSSION THREAD

**Discussion by nooma-42** [zkSync-Community-Hub/zksync-developers#291] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

The comment wrote a sample usage saying that this is for metamask injection. However, this signer inherits from jsonRpcSigner. 

1. Why not wrap abstract class `signer` or `wallet` from vanilla ethers? given metamask injection mentioned above. 
2. Is this comment conflict with the implementation?

```=typescript
// This class is to be used on the frontend, with metamask injection.
// It only contains L2 operations. For L1 operations, see L1Signer.
// Sample usage:
// const provider = new zkweb3.Web3Provider(window.ethereum);
// const signer = provider.getSigner();
// const tx = await signer.sendTransaction({ ... });
export class Signer extends AdapterL2(ethers.providers.JsonRpcSigner) {
```
ref: https://github.com/zksync-sdk/zksync-ethers/blob/754e723e9ff1f63eb46e31171824e39210779652/src/signer.ts#L96

### Expected Behavior

wrap abstract class `signer` or `wallet` from vanilla ethers

### Code Example

```=typescript
// This class is to be used on the frontend, with metamask injection.
// It only contains L2 operations. For L1 operations, see L1Signer.
// Sample usage:
// const provider = new zkweb3.Web3Provider(window.ethereum);
// const signer = provider.getSigner();
// const tx = await signer.sendTransaction({ ... });
export class Signer extends AdapterL2(ethers.providers.JsonRpcSigner) {
```
ref: https://github.com/zksync-sdk/zksync-ethers/blob/754e723e9ff1f63eb46e31171824e39210779652/src/signer.ts#L96

### Repo Link (Optional)

https://github.com/zksync-sdk/zksync-ethers/blob/754e723e9ff1f63eb46e31171824e39210779652/src/signer.ts#L96

**Comment by danijelTxFusion** [comment, marked as answer]
> Is this comment conflict with the implementation?

The code that you are posting refers to `zksync-ethers` based on the` ethers-v5` implementation. The links that you posted refer to the `ethers-v6` implementation. Just to make it clear, the `ethers-v5` implementation is available on the `zksync-ethers` repository's [ethers-v5](https://github.com/zksync-sdk/zksync-ethers/tree/ethers-v5) branch, while the `ethers-v6` implementation is available on the `main` branch.

> Why not wrap abstract class signer or wallet from vanilla ethers? given metamask injection mentioned above.


In `ethers-v5`, `ethers.provider.Web3Provider` returns `JsonRpcProvider` from the [getSigner()](https://github.com/ethers-io/ethers.js/blob/master/packages/providers/src.ts/json-rpc-provider.ts#L488) method (`Web3Provider` extends `JsonRpcProvider`). The same applies to` ethers-v6` with [ethers.BrowserProvider](https://docs.ethers.org/v6/api/providers/#BrowserProvider-getSigner). Therefore, the `Signer` extends the `JsonRpcProvider` to align with the design of ethers. This is why neither `ethers.Wallet` nor `ethers.Signer` are used. Also, `ethers.Signer` does not work with adapters. `AdapterL2` and `AdapterL1` cannot function with abstract classes as they require a constructor from a concrete class.

**Comment by discpenromo** [comment]
The comment in the code snippet seems to be misleading. Since the signer class inherits from JsonRpcSigner, it is not specifically tailored for metamask injection. It would be more appropriate to wrap the abstract class Signer or Wallet from vanilla ethers if the intention is to support metamask injection. This comment appears to conflict with the actual implementation and should be updated for clarity.

**Comment by sayloonkenneth** [comment]
The comment suggests using this class for frontend usage with metamask injection.
 However, it extends from ethers.providers.JsonRpcSigner instead of wrapping a vanilla ethers signer or wallet.
 Consider wrapping abstract class signer or wallet from vanilla ethers for better compatibility and clarity.

**Comment by cbaca9** [comment]
The comment suggests using the Signer class for frontend usage with Metamask injection, but the implementation inherits from JsonRpcSigner instead of wrapping the abstract signer or wallet class from vanilla ethers.

**Comment by segheamulchoidiaz** [comment]
The issue seems to be related to the implementation of the Signer class in the zksync-ethers library. The comment in the code suggests that the class is to be used on the frontend with Metamask injection, but it inherits from JsonRpcSigner instead of wrapping the abstract class Signer or Wallet from vanilla ethers. The expected behavior is to wrap the abstract class Signer or Wallet from vanilla ethers. Without further information or context, it is difficult to provide a more detailed answer.
